File: Java-SpringBoot/BE-01-InMemory/java-21-springboot-4.0/src/main/java/com/flyrank/userservice/exception/GlobalExceptionHandler.java
package com.flyrank.userservice.exception;

import com.flyrank.userservice.dto.ErrorResponse;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.util.stream.Collectors;

/**
 * Controller Advice utilizing Java 21 Features to format exceptions cleanly.
 */
@RestControllerAdvice
public class GlobalExceptionHandler {

    /**
     * Handle 404 User Not Found exception.
     */
    @ExceptionHandler(UserNotFoundException.class)
    public ResponseEntity<ErrorResponse> handleUserNotFound(UserNotFoundException ex) {
        return ResponseEntity
                .status(HttpStatus.NOT_FOUND)
                .body(new ErrorResponse(ex.getMessage()));
    }

    /**
     * Handle 422 Bean Validation errors.
     */
    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ErrorResponse> handleValidationErrors(MethodArgumentNotValidException ex) {
        String detail = ex.getBindingResult().getFieldErrors().stream()
                .map(this::formatFieldError)
                .collect(Collectors.joining("; "));

        return ResponseEntity
                .status(HttpStatus.UNPROCESSABLE_ENTITY)
                .body(new ErrorResponse(detail));
    }

    /**
     * Handle 422 JSON parsing errors using Java 21 Pattern Matching for switch!
     */
    @ExceptionHandler(HttpMessageNotReadableException.class)
    public ResponseEntity<ErrorResponse> handleHttpMessageNotReadable(HttpMessageNotReadableException ex) {
        String message = ex.getMessage();

        // Java 21 Pattern Matching for switch with null check & guard expressions
        String errorDetail = switch (message) {
            case String msg when msg.contains("Unrecognized field") ->
                    "Unrecognized or forbidden field present in JSON payload";
            case String msg when msg.contains("JSON parse error") ->
                    "Malformed JSON syntax payload";
            case null ->
                    "Invalid JSON request body payload";
            default ->
                    "Invalid JSON request body payload";
        };

        return ResponseEntity
                .status(HttpStatus.UNPROCESSABLE_ENTITY)
                .body(new ErrorResponse(errorDetail));
    }

    private String formatFieldError(FieldError fieldError) {
        String field = fieldError.getField();
        String formattedField = field.replaceAll("([a-z])([A-Z])", "$1_$2").toLowerCase();
        return formattedField + ": " + fieldError.getDefaultMessage();
    }
}

File: Java-SpringBoot/BE-01-InMemory/kotlin-2.3-springboot-3.5/src/main/kotlin/com/flyrank/userservice/exception/GlobalExceptionHandler.kt
package com.flyrank.userservice.exception

import com.flyrank.userservice.dto.ErrorResponse
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

/**
 * Controller Advice in idiomatic Kotlin utilizing pattern matching when expressions.
 */
@RestControllerAdvice
class GlobalExceptionHandler {

    @ExceptionHandler(UserNotFoundException::class)
    fun handleUserNotFound(ex: UserNotFoundException): ResponseEntity<ErrorResponse> {
        return ResponseEntity
            .status(HttpStatus.NOT_FOUND)
            .body(ErrorResponse(ex.message ?: "User not found"))
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationErrors(ex: MethodArgumentNotValidException): ResponseEntity<ErrorResponse> {
        val detail = ex.bindingResult.fieldErrors
            .joinToString("; ") { formatFieldError(it) }

        return ResponseEntity
            .status(HttpStatus.UNPROCESSABLE_ENTITY)
            .body(ErrorResponse(detail))
    }

    @ExceptionHandler(HttpMessageNotReadableException::class)
    fun handleHttpMessageNotReadable(ex: HttpMessageNotReadableException): ResponseEntity<ErrorResponse> {
        val message = ex.message ?: ""

        // Kotlin 2.1+ when expression with boolean conditions (guards)
        val errorDetail = when {
            message.contains("Unrecognized field") -> "Unrecognized or forbidden field present in JSON payload"
            message.contains("JSON parse error") -> "Malformed JSON syntax payload"
            else -> "Invalid JSON request body payload"
        }

        return ResponseEntity
            .status(HttpStatus.UNPROCESSABLE_ENTITY)
            .body(ErrorResponse(errorDetail))
    }

    private fun formatFieldError(fieldError: FieldError): String {
        val field = fieldError.field
        val formattedField = field.replace(Regex("([a-z])([A-Z])"), "$1_$2").lowercase()
        return "$formattedField: ${fieldError.defaultMessage}"
    }
}
